src/bpk1.rs

```diff
@@ -131,70 +153,63 @@ where
         let mut result = Vec::<u8>::new();
         let mut writer = Cursor::new(&mut result);
 
-        writer.write(b"BPK1")?;
+        writer.write_all(b"BPK1")?;
         writer.write_u32_le(blocks.len() as u32)?;
-        writer.write_u32_le(7)?;
+        writer.write_u32_le(BPK1_BLOCK_NAME_MAX_LENGTH)?;
 
         let file_size_pos = writer.position();
-        writer.write_u32_le(0)?;
+        writer.write_u32_le(0)?; // file size
+        writer.write_u32_le(0)?; // header size
 
-        let header_size_pos = writer.position();
-        writer.write_u32_le(0)?;
-
-        writer.write(&[0; 0x2c])?; // padding
+        writer.write_all(&[0; 0x2c])?; // padding
 
         let header_start_pos = writer.position();
 
-        const BLOCK_HEADER_SIZE: u8 = 0x4 + 0x4 + 0x4 + 0x8;
-
         for block in &blocks {
             writer.write_u32_le(0)?; // will be offset
             writer.write_u32_le(block.data.len() as u32)?;
             writer.write_u32_le(calc_bpk1_checksum(&block.data))?;
-            writer.write(&string_to_bpk1_bytes(&block.name))?;
+            writer.write_all(&cstring_to_bpk1_bytes(&block.name))?;
         }
 
         let header_size = writer.position() - header_start_pos;
 
-        writer.write_zeroes(0x10 - (header_size % 0x10) as usize)?; // padding
+        let pad = header_size % 0x10;
+        if pad != 0 {
+            writer.write_zeroes((0x10 - pad) as usize)?; // padding
+        }
 
         let data_start_pos = writer.position();
-        writer.set_position(header_size_pos);
-        writer.write_u32_le((data_start_pos - header_start_pos) as u32)?;
-        writer.set_position(data_start_pos);
 
-        let mut index: u8 = 0;
-        for block in &blocks {
+        for (index, block) in (0_u8..).zip(blocks.iter()) {
```

Comment:
Use `.enumerate()` instead, this doesn't need to be a u8. This doesn't save any memory over a usize anyway due to alignment